source/common/http/filter/extauth.cc

```diff
@@ -0,0 +1,308 @@
+#include "common/http/filter/extauth.h"
+
+#include "common/common/assert.h"
+#include "common/common/enum_to_int.h"
+#include "common/http/message_impl.h"
+#include "common/http/utility.h"
+
+namespace Envoy {
+namespace Http {
+
+/**
+ * A pass-through filter that talks to an external authn/authz service.
+ *
+ * When Envoy receives a request for which this filter is enabled, an
+ * asynchronous request with the same HTTP method and headers, but an empty
+ * body, is made to the configured external auth service. The original
+ * request is stalled until the auth request completes.
+ *
+ * If the auth request returns HTTP 200, the original request is allowed
+ * to continue. If any headers are listed in the extauth filter's "headers"
+ * array, those headers will be copied from the auth response into the
+ * original request (overwriting any duplicate headers).
+ *
+ * If the auth request returns anything other than HTTP 200, the original
+ * request is rejected. The full response from the auth service is returned
+ * as the response to the rejected request.
+ *
+ * Note that at present, a call to the external service is made for _every
+ * request_ being routed.
+ */
+
+static LowerCaseString header_to_add(std::string("x-ambassador-calltype"));
+static LowerCaseString value_to_add(std::string("extauth-request"));
+
+ExtAuth::ExtAuth(ExtAuthConfigConstSharedPtr config) : config_(config) {}
```

Comment:
`std::move` the pointer in, especially if this filter is constructed per-request

Reply:
`ExtAuth::ExtAuth(const ExtAuthConfigConstSharedPtr& config) : config_(config) {}` ?

Reply:
That still copies the shared pointer. Passing by value and moving avoids the copy in this ctor. The caller can move their copy into the ctor if possible.